File: Trees, Binary Trees, & Binary Search Trees/SymmetricBinaryTree/SymmetricBinaryTree.java
/*
  This repository is no longer actively maintained. To find all solutions
  to this problem (and practice coding more problems) at:

  ~~~~~~~~~~~~~~~~~~~~~~~~~
  https://backtobackswe.com
  ~~~~~~~~~~~~~~~~~~~~~~~~~
*/

class Solution {
  public boolean isSymmetric(TreeNode root) {
    if (root == null) {
      return true;
    }

    return checkSymmetry(root.left, root.right);
  }

  private boolean checkSymmetry(TreeNode leftSubtreeRoot, TreeNode rightSubtreeRoot) {
    if (leftSubtreeRoot == null && rightSubtreeRoot == null) {
      return true;
    }

    if (leftSubtreeRoot != null && rightSubtreeRoot != null) {
      return leftSubtreeRoot.val == rightSubtreeRoot.val &&
        checkSymmetry(leftSubtreeRoot.right, rightSubtreeRoot.left) &&
        checkSymmetry(leftSubtreeRoot.left, rightSubtreeRoot.right);
    }

    return false;
  }
}

File: Trees, Binary Trees, & Binary Search Trees/SymmetricBinaryTree/SymmetricBinaryTree.py
"""
    Symmetric Tree - LeetCode: https://leetcode.com/problems/symmetric-tree/
    Runtime complexity : O(N)
    Space complexity : O(h)
"""

class Solution(object):
    def isSymmetric(self, root):

        def checker(subtree_0, subtree_1):
            if not subtree_0 and not subtree_1:
                return True
            elif subtree_0 and subtree_1:
                return (subtree_0.val == subtree_1.val and checker(subtree_0.left, subtree_1.right)
                        and checker(subtree_0.right, subtree_1.left))
            return False
        return not root or checker(root.left, root.right)